src/twisted/python/deprecate.py

```diff
@@ -696,6 +696,79 @@ def _passed(argspec, positional, keyword):
 
 
 
+def _passed_signature(signature, positional, keyword):
+    """
+    Take an I{inspect.Signature}, a tuple of positional arguments, and a dict of
+    keyword arguments, and return a mapping of arguments that were actually
+    passed to their passed values.
+
+    @param signature: The signature of the function to inspect.
+    @type signature: I{inspect.Signature}
+
+    @param positional: The positional arguments that were passed.
+    @type positional: L{tuple}
+
+    @param keyword: The keyword arguments that were passed.
+    @type keyword: L{dict}
+
+    @return: A dictionary mapping argument names (those declared in
+        C{signature}) to values that were passed explicitly by the user.
+    @rtype: L{dict} mapping L{str} to L{object}
+    """
+    result = {}
+    kwargs = None
+    for (n, (name, param)) in enumerate(signature.parameters.items()):
+        if param.kind == inspect.Parameter.VAR_POSITIONAL:
+            # Varargs, for example: *args
+            result[name] = positional[n:]
+            n = len(positional)
+        elif param.kind == inspect.Parameter.VAR_KEYWORD:
+            # Variable keyword args, for example: **my_kwargs
+            kwargs = result[name] = {}
+        elif param.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD:
+            if n < len(positional):
+                result[name] = positional[n]
+    if len(positional) > (n + 1):
+        raise TypeError("Too many arguments.")
+    for name, value in keyword.items():
+        if name in signature.parameters.keys():
+            if name in result:
+                raise TypeError("Already passed.")
```

Comment:
This mirrors the behavior of the old code but it sure would be nice to know *what* was already passed.